src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ElementActions.hpp

```diff
@@ -0,0 +1,162 @@
+        [alpha](const gsl::not_null<db::item_type<residual_tag>*> r,
+                const gsl::not_null<db::item_type<fields_tag>*> x,
+                const db::item_type<operand_tag>& p,
+                const db::item_type<operator_tag>& Ap) noexcept {
+          *x += alpha * p;
+          *r -= alpha * Ap;
+        },
+        get<operand_tag>(box), get<operator_tag>(box));
+
+    // Compute new residual norm in a second global reduction
+    const auto& r = get<residual_tag>(box);
+    const double local_residual_magnitude_square = inner_product(r, r);
+
+    Parallel::contribute_to_reduction<UpdateResidual<ParallelComponent>>(
+        local_residual_magnitude_square,
+        Parallel::get_parallel_component<ParallelComponent>(cache)[array_index],
+        Parallel::get_parallel_component<ResidualMonitor<Metavariables>>(cache),
+        CkReduction::sum_double);
+  }
+};
+
+struct UpdateOperand {
+  template <typename... DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent,
+            Requires<tmpl2::flat_any_v<cpp17::is_same_v<
+                db::add_tag_prefix<LinearSolver::Tags::Operand,
+                                   typename Metavariables::system::fields_tag>,
+                DbTags>...>> = nullptr>
+  static auto apply(db::DataBox<tmpl::list<DbTags...>>& box,
+                    tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ArrayIndex& array_index, const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/,
+                    const double res_ratio, const bool terminate) noexcept {
+    using fields_tag = typename Metavariables::system::fields_tag;
+    using operand_tag =
+        db::add_tag_prefix<LinearSolver::Tags::Operand, fields_tag>;
+    using residual_tag =
+        db::add_tag_prefix<LinearSolver::Tags::Residual, fields_tag>;
+
+    // Prepare conjugate gradient for next iteration
+    db::mutate<operand_tag>(
+        make_not_null(&box),
+        [res_ratio](const gsl::not_null<db::item_type<operand_tag>*> p,
+                    const db::item_type<residual_tag>& r) noexcept {
+          *p = r + res_ratio * *p;
+        },
+        get<residual_tag>(box));
+
+    // Increment iteration id
+    db::mutate<LinearSolver::Tags::IterationId,
+               ::Tags::Next<LinearSolver::Tags::IterationId>>(
+        make_not_null(&box), [](const gsl::not_null<IterationId*> iteration_id,
+                                const gsl::not_null<IterationId*>
+                                    next_iteration_id) noexcept {
+          iteration_id->step_number++;
+          next_iteration_id->step_number = iteration_id->step_number + 1;
+        });
+
+    // Terminate when the residual vanishes to machine precision
+    Parallel::get_parallel_component<ParallelComponent>(cache)[array_index]
```

Comment:
I think you can terminate by just having an

```
if (not terminate) {
  get_par_comp<>(cache)[array_index].perform_algorithm();
}
```

but I might be missing something subtle

Reply:
I need to `set_terminate(false)` before the call to `perform_algorithm` since the implementation of `perform_algorithm` doesn't do that. It just returns immediately `if (performing_action_ or get_terminate())`.

Reply:
I'm saying just don't call `perform_algorithm()` when you don't need to.

Reply:
Oh I tried that but it needs the trigger.

Reply:
I don't understand what you mean by `trigger`, could you elaborate?

Reply:
Wait, I think you're relying on `set_terminate` being called before `perform_algorithm`. This is not guaranteed.